### ISSUE TITLE
Quoting on mobile

### ISSUE THREAD

**Issue by b-hodges** [Rapptz/discord.py#1032]
With iOS version 11 the default quote mark is the unicode `“` and `”` rather than ascii `"`. Currently the commands extension doesn't recognize the unicode characters as quotes. It would be nice to let the unicode quotes be used to quote command arguments.

**Comment by Rapptz** [comment]
There is no good way to detect different quotation marks (see: any of `«`, `‹`, `»`, `›`, `„`, `‚`, `“`, `‟`, `‘`, `‛`, `”`, `’`, `"`, `❛`, `❜`, `❟`, `❝`, `❞`, `❮`, `❯`, `〝`, `〞`, `〟`, `＂`) programmatically without hard coding this list. Which I'm still missing a few.

**Comment by mylbp2ps3** [comment]
How would one go about adding this list?

**Comment by b-hodges** [comment]
It looks like the quoting occurs in this function: https://github.com/Rapptz/discord.py/blob/rewrite/discord/ext/commands/view.py#L111
Is this something that should be changed, or not?

**Comment by mylbp2ps3** [comment]
I'd say changing line 117, could make it work. I'll try it out.

**Comment by b-hodges** [comment]
117 won't be enough, that's only the opening quote. There's more quotes at 140 and 150. Plus the ending quote should match the type of the opening quote.

**Comment by mylbp2ps3** [comment]
So, I've added a quotes array, and simply replaced line 117 with:

```
is_quoted = False
if current in quotes:
    is_quoted = True
```
140 I changed to:
`if next_char in quotes:`
150 to:
`if current in quotes:`

This seems to work pretty well, though haven't tested it intensively, but it works! Right now the ending quote doesn't need to match the opening quote. Though it is doable, but a bit complicated. Since some quotes aren't similar, but belong together. For example. "❝", "❞".

But it could be solved with a dictionary. Something like:
```
    quotes = {
        "«": "»",
        "❛": "❜",
        '"': '"'
    }
```
So the keys are the opening quotes, and the corresponding item is the closing quote. We could then store the used opening quote's closing quote in a variable, and use that to check against when checking for a closing quote.

**Comment by b-hodges** [comment]
I've added pull request #1035 to potentially add handling for quotes. I have quotes paired off in a dictionary like mylbp2ps3 suggested, but didn't add all of them.

**Comment by mylbp2ps3** [comment]
I believe I've now got matching quotes working quite well. Went over all the unicode characters marked with `Quotation Mark=Yes` (30 in total) and paired them together. Had to make the values in the dictionary into a list, since some had two kinds of closing quotes. Final result in full is this:
```
def quoted_word(view):
    quotes = {
        '"': ['"'],
        "'": ["'"],
        "«": ["»"],
        "‘": ["’"],
        "‚": ["‛"],
        "“": ["”"],
        "„": ["‟"],
        "‹": ["›"],
        "⹂": ["⹂"],
        "「": ["」"],
        "『": ["』"],
        "〝": ["〞", "〟"],
        "﹁": ["﹂"],
        "﹃": ["﹄"],
        "＂": ["＂"],
        "＇": ["＇"],
        "｢": ["｣"]
    }
    current = view.current

    if current is None:
        return None

    is_quoted = False
    if current in quotes.keys():
        opening_quote = current
        is_quoted = True
    result = [] if is_quoted else [current]

    while not view.eof:
        current = view.get()
        if not current:
            if is_quoted:
                # unexpected EOF
                raise BadArgument('Expected closing "')
            return ''.join(result)

        # currently we accept strings in the format of "hello world"
        # to embed a quote inside the string you must escape it: "a \"world\""
        if current == '\\':
            next_char = view.get()
            if not next_char:
                # string ends with \ and no character after it
                if is_quoted:
                    # if we're quoted then we're expecting a closing quote
                    raise BadArgument('Expected closing "')
                # if we aren't then we just let it through
                return ''.join(result)

            if next_char in quotes[opening_quote]:
                # escaped quote
                result.append('"')
            else:
                # different escape character, ignore it
                view.undo()
                result.append(current)
            continue

        # closing quote
        if current in quotes[opening_quote]:
            next_char = view.get()
            valid_eof = not next_char or next_char.isspace()
            if is_quoted:
                if not valid_eof:
                    raise BadArgument('Expected space after closing quotation')

                # we're quoted so it's okay
                return ''.join(result)
            else:
                # we aren't quoted
                raise BadArgument('Unexpected quote mark in non-quoted string')

        if current.isspace() and not is_quoted:
            # end of word found
            return ''.join(result)

        result.append(current)
```

**Comment by Rapptz** [comment]
I merged #1035 so this is fixed now.
